File: data_structure/binary_tree_inorder_tra/recrusive.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:  # the recrusive method
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result=[]
        self.helper(root,result)
        return result
    
    def helper(self,root,result):
        if root:
            self.helper(root.left,result)
            result.append(root.val)
            self.helper(root.right,result)

File: data_structure/binary_tree_inorder_tra/recrusive.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> node;
        inorder(root,node);
        return node;
    }
    void inorder(TreeNode* root,vector<int>& node){
        if(root){
            inorder(root->left,node);
            node.push_back(root -> val);
            inorder(root->right,node);
        }
    }
};

/*
int a = b is setting a's VALUE to b's VALUE
int* a = &b is setting a's VALUE to the ADDRESS of b
int& a = b is setting a's ADDRESS to b's ADDRESS (a is a reference to b)

*/